### PR TITLE
[composable-controller] Better typing for state, messenger, strict type checks for inputs, remove `#controllers` field

### LINKED ISSUE #3716: `composable-controller`: Replace use of `any` with proper types (non-test files only)
Look for the following string in non-test files to find these: ``// TODO: Replace `any` with type``

At time of writing, there were 2 instances.

### LINKED ISSUE #3907: [composable-controller] Fix incorrect `#updateChildController` behavior and improve typing
There are several issues with the current ComposableController implementation that should be addressed before further updates are made to the controller (e.g. #3627).

- Remove `#controllers` class field, which is not being updated by `#updateChildController` or anywhere else.
  - Removing this makes it clear that the list of child controllers to be composed is determined at class instantiation and cannot be altered later. 
  - This behavior is consistent with `#updateChildController` being a private method.
- Type BaseController, ComposableController state with `Record<string, Json>`
- Add a `isBaseController` type guard. 
  - Remove the deprecated `subscribed` property from `BaseController`.

### PR DESCRIPTION
**Pull request by MajorLift**
## Explanation

There are several issues with the current ComposableController implementation that should be addressed before further updates are made to the controller (e.g. #3627).

- Removes `#controllers` class field, which is not being updated by `#updateChildController` or anywhere else.
  - Removing this makes it clear that the list of child controllers to be composed is determined at class instantiation and cannot be altered later. 
  - This behavior is consistent with `#updateChildController` being a private method.
- Types BaseController, ComposableController state with `Record<string, Json>`
  - Opted to use `any` to disable BaseController state type constraint, as there is no straightforward way to type `BaseControllerV1` state to be compatible with `Json`.
- Adds a `isBaseController` type guard, ~and removes the deprecated `subscribed` property from `BaseController`~.
- Removes `ControllerList` type in anticipation of https://github.com/MetaMask/core/issues/3627. Internally, `ControllerInstance` will be used to type child controllers or unions and tuples thereof.
- Adds `BaseControllerV{1,2}Instance` types.
- Populates state metadata object in constructor with child controllers.

## References

- Blocks #3627 
- Closes #3716
  - Replaces some TODOs with comment explaining necessity of `any` usage.
- Closes #3907

## Changelog

Recorded under "Unreleased" heading in CHANGELOG files.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/composable-controller/tsconfig.json
Comment (MajorLift) [inline packages/composable-controller/tsconfig.json:12]: Added `json-rpc-engine` as a TypeScript project reference (but not in the build tsconfig) instead of introducing it as a dependency, since it's only being used in a test file.

File: packages/composable-controller/src/ComposableController.ts
Comment (MajorLift) [inline packages/composable-controller/src/ComposableController.ts:71]: This doesn't interfere with `childState` types being correctly inferred in `#updateChildController`

File: packages/composable-controller/src/ComposableController.ts
Comment (MajorLift) [inline packages/composable-controller/src/ComposableController.ts:74]: Previous typing of `ControllerInstance['state']` implicitly evaluated as `any`. 

This is the error if `any` is replaced with something wider than `Json` e.g. `unknown`:
```ts
Type 'ComposableControllerState' does not satisfy the constraint 'Record<string, Json>'.
  'string' index signatures are incompatible.
    Type 'Record<string, unknown>' is not assignable to type 'Json'.
```

File: packages/composable-controller/src/ComposableController.ts
Comment (MajorLift) [inline packages/composable-controller/src/ComposableController.ts:79]: Ideally, this would be something that covers the state types of both BaseController versions:
```ts
Record<string, BaseState & Record<string, unknown> | Record<string, Json>>
```
But this results in an implicit any error, and it conflicts with the messenger typing for allowed events. 

This seems to be the best possible type until `AllowedEvents` is narrowed by #3627.

File: packages/composable-controller/src/ComposableController.ts
Comment (MajorLift) [inline packages/composable-controller/src/ComposableController.ts:95]: No reason to keep this as `string` since ComposableController doesn't use any actions from other controllers.

File: packages/composable-controller/src/ComposableController.ts
Comment (MajorLift) [inline packages/composable-controller/src/ComposableController.ts:168]: Test case added for this branch.

File: packages/composable-controller/src/ComposableController.ts
Comment (MajorLift) [inline packages/composable-controller/src/ComposableController.ts:159]: Since the child controller is validated as BaseControllerV2, its state type is guaranteed to be `Record<string, Json>`. It's therefore safe to narrow `childState` from `Record<string, unknown>`.

File: packages/composable-controller/src/ComposableController.ts
Comment (MajorLift) [inline packages/composable-controller/src/ComposableController.ts:67]: Added redundant property checks for extra runtime safety.

File: packages/composable-controller/src/ComposableController.ts
Comment (MajorLift) [inline packages/composable-controller/src/ComposableController.ts:38]: The BaseControllerV1 `childState` in `#updateChildController` is correctly inferred to this state type.

File: packages/composable-controller/src/ComposableController.ts
Comment (MajorLift) [inline packages/composable-controller/src/ComposableController.ts:66]: Alternatively, we could keep the `#controllers` field and add something like the following to `#updateChildController`.

```ts
if (!this.state[name]) {
  this.#controllers.push(controller)
}
```

I prefer removing the field because it makes it unambiguous that child controllers aren't meant to be added, and their state and subscriptions aren't meant to be mutated after instantiation. 

Also, having ComposableController depend on mutable internal state would introduce complications for #3627.

File: packages/composable-controller/tsconfig.json
Comment (mcmire) [reply packages/composable-controller/tsconfig.json:12]: Do we also want to add `json-rpc-engine` as a dev dependency? Maybe it makes no difference from a TypeScript perspective, but it communicates the dependency from a Yarn/NPM perspective.

File: packages/base-controller/src/BaseControllerV2.ts
Comment (mcmire) [inline packages/base-controller/src/BaseControllerV2.ts:116]: @Gudahtt I know that we tried to remove this once, and you previously reverted it because we wanted to avoid another major `base-controller` release. What are your thoughts on this?

**Review by mcmire — COMMENTED** [review]
Had a couple of comments, but overall these seem like nice changes.

File: packages/composable-controller/tsconfig.json
Comment (MajorLift) [reply packages/composable-controller/tsconfig.json:12]: Fixed here: https://github.com/MetaMask/core/pull/3904/commits/5da089327847068f62dd1ec971b5a3299bff8eb9

Will a `json-rpc-engine` project reference need to be added to `tsconfig.build.json` as well?

Comment (mcmire) [reply packages/composable-controller/tsconfig.json:12]: If it's only being used for tests, then you're correct, I don't think adding it to `tsconfig.build.json` would be needed. You should be able to test this by running `yarn build` — you shouldn't see any errors.

Comment (MajorLift) [reply packages/composable-controller/tsconfig.json:12]: No errors on yarn build! Think we're good to go.

File: packages/base-controller/src/BaseControllerV2.ts
Comment (MajorLift) [reply packages/base-controller/src/BaseControllerV2.ts:116]: It looks like we could piggy-back on the major release that will follow https://github.com/MetaMask/core/pull/3611.

Comment (mcmire) [reply packages/base-controller/src/BaseControllerV2.ts:116]: It looks like #3611 is blocked by #3908, which is blocked by the TransactionController multichain PR. So once that's merged and we get a new release of TransactionController and SelectedNetworkController we can return to this.

Comment (MajorLift) [reply packages/base-controller/src/BaseControllerV2.ts:116]: As per our discussion, there should be no issue with merging this PR first and keeping it unreleased, as long as it's eventually batched into the same release as #3611. The contents of this PR and #3908 are not dependent on each other, so there's no need to wait for it to merge first.

Note from @mcmire: Until #3611 is ready for release, we will need to monitor release candidates to make sure `@metamask/composable-controller` is not included.

Comment (MajorLift) [reply packages/base-controller/src/BaseControllerV2.ts]: Removed in https://github.com/MetaMask/core/pull/3904/commits/1107e251e8818104ed79e564bd86282d4c522eaf. To be reapplied after https://github.com/MetaMask/core/pull/3611 is merged.

(Changes to base-controller may be disruptive for releases due to https://github.com/MetaMask/create-release-branch/issues/91)

File: packages/base-controller/CHANGELOG.md
Comment (MajorLift) [inline packages/base-controller/CHANGELOG.md:14]:
```suggestion
```

File: packages/composable-controller/src/ComposableController.ts
Comment (mcmire) [inline packages/composable-controller/src/ComposableController.ts:14]: This JSDoc will only appear for this particular type (whether you hover over it or view it via TypeDoc), so saying "the following" would be confusing. Perhaps we ought to add JSDocs for the other two types too?

File: packages/composable-controller/CHANGELOG.md
Comment (mcmire) [inline packages/composable-controller/CHANGELOG.md:21]: Maybe we want to use "Throw" instead of "Throws" for consistency with other entries? Or we could flip it around. Also, this seems breaking to me instead of the `@metamask/utils` addition. Finally, the modification to `AllowedActions` seems like a change rather than a removal, so should it be here too?

``` suggestion
- **BREAKING:** Passing a non-controller into `controllers` constructor option now throws an error ([#3904](https://github.com/MetaMask/core/pull/3904))
- **BREAKING:** The `AllowedActions` parameter of the `ComposableControllerMessenger` type is narrowed from `string` to `never`, as `ComposableController` does not use any external controller actions. ([#3904](https://github.com/MetaMask/core/pull/3904))
- Add `@metamask/utils` ^8.3.0 as a dependency. ([#3904](https://github.com/MetaMask/core/pull/3904))

### Removed

File: packages/composable-controller/src/ComposableController.test.ts
Comment (mcmire) [inline packages/composable-controller/src/ComposableController.test.ts:13]: Nit: We have Jest and TypeScript set up so that if you use `@metamask/*` it will look for an appropriate package in `packages/`. What are your thoughts on:

``` suggestion
import { JsonRpcEngine } from '@metamask/json-rpc-engine';

**Review by mcmire — COMMENTED** [review]
Left some comments. Most are minor. Thanks for removing the BaseControllerV2 change. I think we're close to merging this now!

File: packages/composable-controller/src/ComposableController.ts
Comment (mcmire) [reply packages/composable-controller/src/ComposableController.ts]: Good point, I missed that 😅

**Review by mcmire — APPROVED** [review]
Looks good!

**Pull request merged**
